My Site Dashboard (Phase 2): Quick Start Initial Structure

momo-ozawa opened the pull request:
Part of #17877

## Description
- Adds the initial structure for the Quick Start dashboard card
- Triggers Quick Start from the dashboard view
- Displays a site menu waypoint if the tour is initiated from the dashboard view (if needed)

<img src="https://user-images.githubusercontent.com/6711616/155394130-c2a640a6-eb1a-44fc-a402-85cfc3fedb3a.png" width=200>

## Notes
- TODO: Refine Quick Start dashboard card design
- TODO: Show Quick Start on Site Menu or Dashboard (one or the other, not both)

## How to test

Make sure the MSD feature flag is enabled

1. First we need to reset Quick Start -- Remove "Next steps" from your site menu view if you have it showing. Otherwise, continue to the next step
2. Enable Quick Start from App Settings > Debug 
3. Tap "Home" on the segmented control to display the dashboard
4. ✅ A quick start card dashboard should be displayed
5. Go through all tours on both the "Customize Your Site" and the "Grow Your Audience" checklist
    -  ✅ Make sure that the tour scrolls to the correct element and/or shows a waypoint on the correct element
    -  Also, check the following cases for whether the site menu waypoint is displayed or not:

Quick Start Tour | Tour triggered from dashboard -> Shows site menu waypoint? | Tour triggered from site menu -> Shows site menu waypoint?
-- | -- | -- 
Check your site title | ✖️ | ✖️
Check your site icon | ✖️ | ✖️
Edit your home page | ✅ | ✖️
Review site pages | ✅ | ✖️
View your site| ✅ | ✖️
Social sharing | ✅ | ✖️
Publish a post | ✅ | ✖️
Follow other sites | ✖️ | ✖️
Check your site stats | ✅ | ✖️

https://user-images.githubusercontent.com/6711616/155394382-281921ad-7a24-4a68-850c-2ebc3153f098.mp4

## Regression Notes
1. Potential unintended areas of impact
- Triggering quick start tours from the Quick Start cell in BlogDetailsViewController (i.e. the site menu)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Tested manually, checked every single quick start tour from BlogDetailsViewController

3. What automated tests I added (or what prevented me from doing so)
- n/a
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Add: set up quick start for MySiteViewController

- Add a spotlight view on the segmented control
- Observe quick start tour element changed notification

Commit:
Add: quick start site menu waypoint

Commit:
Refactor: reorganize site picker related files

- Move site picker related files from My Site folder to Site Picker folder

Commit:
Add: initial structure for quick start card cell

Commit:
Refactor: make waypoints settable

- Change waypoints to be settable so that a site menu waypoint can be inserted into the tour as needed

Commit:
Add: shownInBlogDetails property to quick start tour

The shownInBlogDetails property is a way to differentiate between quick start tours that are shown within BlogDetailsViewController or not.

If false, we don’t want to insert a site menu waypoint into the tour.

Commit:
Refactor: remove embedded child vc from stack view

Commit:
Refactor: insert site menu waypoint into tour if needed

Commit:
Refactor: visit site menu waypoint and show next step with delay

Commit:
Refactor: check if current tour origin is blog dashboard or not

momo-ozawa commented:
@pachlava Can you hold off testing this one? I'd like to address the TODOs before I ask you for a review 🙏

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift:
```diff
@@ -9,6 +9,7 @@ import Foundation
 /// identifier on the backend.
 enum DashboardCard: String, CaseIterable {
     case quickActions
+    case quickStart
```
Not necessarily to be done in this PR but you might want to add a property here named `shouldShow`. Then, the `BlogService` will check if this card should really be shown (default is always `true`).

But for this cell, we want to check if there's a quick start enabled in progress or not.

momo-ozawa replied:
Unfortunately enum properties are not settable :(

I think I'll have to inject a Blog into `BlogDashboardService`, so I can call `QuickStartTourGuide.shouldShowChecklist(for: blog)` to check whether quick start is enabled or not.

leandroalonso replied:
Oh, got it! I was thinking of something like:

```swift
    var shouldShow: Bool {
        switch self {
        case .quickStart:
            return QuickStartTourGuide.shouldShowChecklist(for: blog)
        default:
            return true
        }
    }
```

But I see that we would have to inject the `blog` property, that's an interesting challenged, indeed.

momo-ozawa replied:
I ended up taking the approach I mentioned above. a760e51

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController+QuickStart.swift:
```diff
@@ -0,0 +1,23 @@
+extension MySiteViewController {
+
+    func startObservingQuickStart() {
+        observer = NotificationCenter.default.addObserver(forName: .QuickStartTourElementChangedNotification, object: nil, queue: nil) { [weak self] (notification) in
```
Do we really need to store that on an `observer` var?

momo-ozawa replied:
Fixed! d83f15a

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:
```diff
@@ -623,7 +654,7 @@ class MySiteViewController: UIViewController, NoResultsViewHost {
         }
 
-        remove(blogDashboardViewController)
+        removeChildFromStackView(blogDashboardViewController)
```
By doing this, you're forcing this VC to be deleted/regenerated every time the user switches between them.

This in turn, causes the collection view on the dashboard to always recreate itself (including cells, child VCs, etc) — the same for the blog details (it is recreated every time, and it will perform the `preloadMetadata` every time, which is a heavy API call). Is that done on purpose? If so, why?

momo-ozawa replied:
Actually, BlogDetailsVC / BlogDashboardVC isn't recreated every time we switch. `viewDidLoad` is triggered only once for each child vc. You can add a breakpoint in viewDidLoad / deinit to verify this.`viewDidAppear`, however, is called every time we switch — which is expected behavior.

The only difference between `remove(blogDashboardViewController)` and `removeChildFromStackView(blogDashboardViewController)` is that the latter explicitly removes the child VC view from the stackview's arranged subviews, but this doesn't destroy the child VC's

leandroalonso reviewed (commented):
@momo-ozawa I had two issues:

At one moment, the bottom dialog suggested to do the "Create your site" tour (which wasn't crossed). After I accepted, I ended up in this state:

<img src="https://user-images.githubusercontent.com/7040243/155567538-8282440a-8dee-41da-a21a-297917bf3f48.png" width="300">

Also, when I switch to the "Site menu" I can see that the scroll and the dot animation is performed multiple times:

https://user-images.githubusercontent.com/7040243/155567666-3b866623-5632-4789-af48-4b82946cd886.mp4

Commit:
Fix: don’t store observer

Commit:
Refactor: add quick start card to snapshot as needed

Commit:
Refactor: increase delay for next step

momo-ozawa commented:
@leandroalonso Ready for a re-review!

> At one moment, the bottom dialog suggested to do the "Create your site" tour (which wasn't crossed). After I accepted, I ended up in this state:

I think I know why this was happening. If you go to the Dashboard view without enabling QuickStart, we never call `QuickStartTourGuide.setup`. This is probably why the "Create your site" tour wasn't crossed out / marked as completed.

Anyway, a760e51 should resolve this issue, since we're now showing QuickStart only if it's enabled.

> Also, when I switch to the "Site menu" I can see that the scroll and the dot animation is performed multiple times:

I think 03e1349 should resolve this. The reason the animation was being performed multiple times is because the cell is configured multiple times after `viewDidAppear`. I increased the delay from .5 seconds to 1 second to avoid this. Lmk if you have a better idea on how to resolve this issue.


https://user-images.githubusercontent.com/6711616/155709648-5b67b759-a8f7-4644-91c4-08437ecc7686.mp4

Commit:
Merge branch 'trunk' into task/17877-msd-quick-start-initial-structure

# Conflicts:
#	WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift
#	WordPress/WordPress.xcodeproj/project.pbxproj

Commit:
Refactor: adjust safe area insets for quick start

Commit:
Refactor: don’t show site menu waypoint for publish tour

Commit:
Fix: differentiate Card Model based on api response

Otherwise, local cards (i.e. Quick Start / Quick Actions) aren’t distinguishable since the apiResponseDictionary is always nil

Commit:
Merge branch 'trunk' into task/17877-msd-quick-start-initial-structure

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/DashboardCardModel.swift:
```diff
@@ -15,7 +15,8 @@ struct DashboardCardModel: Hashable {
 
     static func == (lhs: DashboardCardModel, rhs: DashboardCardModel) -> Bool {
-        lhs.apiResponseDictionary == rhs.apiResponseDictionary
+        return lhs.id == rhs.id &&
```
```suggestion
       lhs.id == rhs.id &&
```

momo-ozawa replied:
Fixed! 1273a3d

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:
```diff
@@ -2,10 +2,13 @@ import Foundation
 
 class BlogDashboardService {
+
+    private let blog: Blog
```
The `DashboardService` persists even if you change blogs.

The way it is now, the `blog` property is set when this is initialized and it never changes. You can test it:

1. Open the dashboard for a blog with quick start enabled
2. Change to another blog
3. Quick start is still displayed

You may want to change `fetch` to accept a `Blog` instead of a `wpComID`.

momo-ozawa replied:
Fixed! dba2bd5

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:
```diff
@@ -64,6 +67,12 @@ private extension BlogDashboardService {
             } else {
+
+                // Don't add the Quick Start card if we shouldn't show Quick Start
+                if card == .quickStart && !QuickStartTourGuide.shouldShowChecklist(for: blog) {
```
I think we can follow the open-closed principle here by moving this logic to the `DashboardCard` enum.

You can add a function on the `DashboardCard` like this:

```
    func shouldShow(for blog: Blog) -> Bool {
        switch self {
        case .quickActions:
            return true
        case .quickStart:
            return QuickStartTourGuide.shouldShowChecklist(for: blog)
        case .posts:
            return true
        case .todaysStats:
            return true
        }
    }
```

Then here you can just:

```swift
                if card.shouldShow(for: blog) {
                    let section = DashboardCardSection(id: card)
                    let item = DashboardCardModel(id: card)

                    snapshot.appendSections([section])
                    snapshot.appendItems([item], toSection: section)
                }
```

This way you can control that on a card-basis without the need for changing the service.

Let's say in the future we want to hide a card (a promotional card, for example), we can just change this on the enum and done. :)

momo-ozawa replied:
Fixed! 24366c6

This is a much better solution, I forgot about the option of adding methods to enums 🙈

leandroalonso commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController.swift:
```diff
@@ -228,7 +244,10 @@ class MySiteViewController: UIViewController, NoResultsViewHost {
-            segmentedControl.bottomAnchor.constraint(equalTo: segmentedControlContainerView.bottomAnchor)
+            segmentedControl.bottomAnchor.constraint(equalTo: segmentedControlContainerView.bottomAnchor),
+            segmentedControl.heightAnchor.constraint(equalToConstant: Constants.segmentedControlHeight),
+            siteMenuSpotlightView.leadingAnchor.constraint(equalTo: segmentedControl.leadingAnchor, constant: -Constants.siteMenuSpotlightOffset),
```
```suggestion
            
            // Quick start spotlight
            siteMenuSpotlightView.leadingAnchor.constraint(equalTo: segmentedControl.leadingAnchor, constant: -Constants.siteMenuSpotlightOffset),
```

The number of constraints are getting bigger, this can help to understand what's what (or maybe extracting to different methods).

momo-ozawa replied:
Fixed! 418cf60

leandroalonso reviewed (commented):
@momo-ozawa tested the flows as it follows:

1. With the MSD flag on, in both tabs, all good ✅
2. With the flag disabled, to make sure the current flow is still the same ✅

Left two improvements suggestions, one of them it's a bug I think (the `blog` not changing).

Commit:
Fix: set service blog when view model blog gets set

Commit:
Refactor: add shouldShow method for dashboard cards

Commit:
Refactor: omit return keyword

Commit:
Refactor: reorganize constraints for clarity

Commit:
Fix: blog dashboard tests

Commit:
Refactor: pass blog to fetch methods instead of dotComID

Commit:
Fix: set context to nil on teardown

leandroalonso committed:
test: rename the test class name to avoid failures
